include/uk/event.h

```diff
@@ -0,0 +1,296 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __UK_EVENT_H__
+#define __UK_EVENT_H__
+
+/**
+ * ## Overview
+ *
+ * `uk_event` decouples the definition of an event and the definition of
+ * callback routines that should be called whenever the event ocurrs. This way
+ * a library (A) can provide an event and a library (B) can register handlers
+ * for this event without creating a dependency (A)->(B). Handler registration
+ * is done at link time by collecting all handler functions for a particular
+ * event and putting their addresses into an event-specific function table.
+ * Handlers can indicate that they have handled the event, in which case no
+ * other handlers are called, if desired. Handlers can also be assigned a
+ * priority class to enforce a certain execution order.
+ *
+ * ## Using Events
+ *
+ * In the following, we define an event `myevent` in library (A). We encapsulate
+ * additional event-specific information in `struct myevent_data`. Library (B)
+ * includes the declaration of `struct myevent_data` and registers three
+ * handler functions, where `myhandler1` and `myhandler2` allow other handlers
+ * to be called afterwards and `myhandler3` terminates handler invocation.
+ * Note: Since there are no guarantees in which order handlers of the same
+ * priority class may be called, it is only guaranteed that `handler1` will be
+ * the first of libB's handlers that is invoked. However, `handler2` may or may
+ * not be called depending on if `handler3` is executed before.
+ *
+ * **libA/include/uk/myevent.h**:
+ * ```C
+ * struct myevent_data {
+ *       int dummy;
+ * };
+ * ```
+ * **libA/myevent.c:**
+ * ```C
+ * #include <uk/event.h>
+ * #include <uk/myevent.h>
+ *
+ * UK_EVENT(myevent);
+ *
+ * void some_function(void)
+ * {
+ *       struct myevent_data = { .dummy = 42 };
+ *       int rc;
+ *
+ *       rc = uk_raise_event(myevent, &myevent_data);
+ *       if (!rc)
+ *               uk_pr_err("myevent not handled!\n");
+ * }
+ * ```
+ * **libB/myhandler.c:**
+ * ```C
+ * #include <uk/event.h>
+ * #include <uk/myevent.h>
+ *
+ * int handler1(void *arg)
+ * {
+ *       struct myevent_data *data = (struct myevent_data *)arg;
+ *       ...
+ *       return UK_EVENT_HANDLED_CONT;
+ * }
+ *
+ * int handler2(void *arg)
+ * {
+ *       return UK_EVENT_HANDLED_CONT;
+ * }
+ *
+ * int handler3(void *arg)
+ * {
+ *       return UK_EVENT_HANDLED;
+ * }
+ *
+ * UK_EVENT_HANDLER_PRIO(handler1, UK_PRIO_EARLIEST);
+ * UK_EVENT_HANDLER_PRIO(handler2, UK_PRIO_LATEST);
+ * UK_EVENT_HANDLER_PRIO(handler3, UK_PRIO_LATEST);
+ * ```
+ */
+
+#include <uk/essentials.h>
+#include <uk/prio.h>
+
+#ifdef CONFIG_LIBUKDEBUG
+#include <uk/assert.h>
+#endif /* CONFIG_LIBUKDEBUG */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifdef CONFIG_LIBUKDEBUG
+#define __event_assert(x) UK_ASSERT(x)
```

Comment:
I stole this from `/include/uk/refcount.h`, where it does not use the `uk` prefix. I have changed it anyways 😃

But we should have a discussion about this because I heavily dislike `UK_ASSERT` not being part of the generic headers and having to check for `ukdebug` all the time. This is especially stupid as we add dependencies for `ukdebug` everywhere, also in the platform code. If we were stringent, we would have to add checks everywhere also for the presence of the `uk_pr_*` functions or simply accept that you cannot build without `ukdebug`, which would make such checks obsolete. I already had a short discussion about this with @michpappas here: https://github.com/unikraft/unikraft/pull/338#discussion_r829961496

Reply:
Good points. I would now also say that we should change it to refcount, too. In general, I agree with you, `UK_ASSERT` became such an important macro for us. Together with the compile-time assertions, it might make sense to move it out from `ukdebug` and place it under the global `/include/uk` folder. However, the problem that we have is that the assertion currenlty depend on `ukdebug's` printing system :thinking:... this is why they stick with `ukdebug` currently. In general we have only problems with headers that are in `/include` which we explicitly do not want to imply any library dependency so that we keep highest flexibility. Maybe we can do an exception for assertions to let print messages only if `UKDEBUG` is available... similar to this assertion availability helper here. What do you think?